### DISCUSSION TITLE
How to center text with draw?

### DISCUSSION THREAD

**Discussion by printmaps** [ImageMagick/ImageMagick#4444] [Help] [answered]
I have something like this ...

```
convert -size "600 x 150" canvas:gray slot-600x150.png
convert slot-600x150.png -fill white -draw "circle 100,75 102,75" -draw "circle 300,75 302,75" -draw "circle 500,75 502,75" marked-slot-600x150.png
```
![marked-slot-600x150](https://user-images.githubusercontent.com/26689799/140511746-00d0e296-b073-4045-99bb-c95938f62662.png)

... and want to center arbitrary text on the white points.

![text-marked-slot-600x150](https://user-images.githubusercontent.com/26689799/140511811-94a19998-9cbf-4cfd-8050-3d5e2e2a7377.png)

How to archive this with the draw option for text?

**Comment by snibgo** [comment]
What version of IM, on what platform? I assume IM v7 on Windows.

I would not use "-draw" for the text. Instead:
```
magick ^
  -size 600x150 canvas:gray -fill white ^
  -draw "circle 100,75 102,75" ^
  -draw "circle 300,75 302,75" ^
  -draw "circle 500,75 502,75" ^
  ( -fill Black -background None +size label:AAABBCC ^
    -geometry +%%[fx:100-w/2]+%%[fx:75-h-2] ) ^
  -composite ^
  ( -fill Black -background None +size label:DDEEFF ^
    -geometry +%%[fx:300-w/2]+%%[fx:75-h-2] ) ^
  -composite ^
  ( -fill Black -background None +size label:GGGGHHHHIIIIJJJJ ^
    -geometry +%%[fx:502-w/2]+%%[fx:75-h-2] ) ^
  -composite ^
  out.png
```

Note that:
```
-geometry +%%[fx:100]+%%[fx:75]
```
... would put the top-left of the text at loction (100,75). We move it left by half the width of the text, and move it up by the height of the text and another 2 pixels.
```
-geometry +%%[fx:100-w/2]+%%[fx:75-h-2]
```
![out](https://user-images.githubusercontent.com/33812368/140519315-6e01dbfe-313c-4f45-82d7-5a2222f19659.png)

**Comment by printmaps** [comment]
I'm using IM7 on macOS (7.1.0-13) oder Ubuntu with zsh or bash. Unfortunately I can make it work there. Probably problems with shell expansion or interpretation.

**Comment by snibgo** [comment]
If you showed your bash command, I could probably tell you what was wrong with it. This works for me:
```
magick \
  -size 600x150 canvas:gray -fill white \
  -draw "circle 100,75 102,75" \
  -draw "circle 300,75 302,75" \
  -draw "circle 500,75 502,75" \
  \( -fill Black -background None +size label:AAABBCC \
    -geometry +%[fx:100-w/2]+%[fx:75-h-2] \) \
  -composite \
  \( -fill Black -background None +size label:DDEEFF \
    -geometry +%[fx:300-w/2]+%[fx:75-h-2] \) \
  -composite \
  \( -fill Black -background None +size label:GGGGHHHHIIIIJJJJ \
    -geometry +%[fx:502-w/2]+%[fx:75-h-2] \) \
  -composite \
  out.png
```

**Comment by printmaps** [comment]
> If you showed your bash command, I could probably tell you what was wrong with it.

My shell problem was escaping the parentheses: '(' -> '\('. Thanks for pointing to this. With two small changes (font added, geometry parameter in ") it works for me on macOS (zsh) and Ubuntu (bash):

```
magick \
  -size 600x150 canvas:gray -fill white \
  -draw "circle 100,75 102,75" \
  -draw "circle 300,75 302,75" \
  -draw "circle 500,75 502,75" \
  -font DejaVu-Sans-Oblique \
  -pointsize 22 \
  \( -fill Black -background None +size label:AAABBCC \
    -geometry "+%[fx:100-w/2]+%[fx:75-h-2]" \) \
  -composite \
  \( -fill Black -background None +size label:DDEEFF \
    -geometry "+%[fx:300-w/2]+%[fx:75-h-2]" \) \
  -composite \
  \( -fill Black -background None +size label:GGGGHHHHIIIIJJJJ \
    -geometry "+%[fx:502-w/2]+%[fx:75-h-2]" \) \
  -composite \
  out.png
```

I need this kind of labeling for huge images (eg. 10000x10000 pixel). I will run some tests, in order to evaluate that the performance of the recommended strategy is sufficient. Further I need to rotate (flip) the labels (180°). I will come back to this discussion when I have the test results.

Basically my question is answered. Thank's for looking into this and pointing me to the right direction.

**Comment by printmaps** [comment, marked as answer]
That's my 'final' solution ... a 'complex' magick call:

```
magick \
    -size "$DruckBreitePixel"x"$DruckHoehePixel" canvas:white \
    \( -colorspace sRGB "$DateinamePlasma" \) -gravity Center -composite \
    \( -colorspace sRGB "$DateinameObjekt" \) -gravity Center -composite \
    -font DejaVu-Sans-Oblique \
    -pointsize 44 \
    \( -fill Black -background None +size "label:$BeschriftungObenLinks" -geometry "+%[fx: $BeschriftungMitteLinkesDrittel - w / 2] +%[fx: $BeschriftungOffsetVertikalPixel - h / 2]" -rotate 180 \) -gravity NorthWest -composite \
    \( -fill Black -background None +size "label:$BeschriftungObenMitte" -geometry "+%[fx: $BeschriftungMitteMittleresDrittel - w / 2] +%[fx: $BeschriftungOffsetVertikalPixel - h / 2]" -rotate 180 \) -gravity NorthWest -composite \
    \( -fill Black -background None +size "label:$BeschriftungObenRechts" -geometry "+%[fx: $BeschriftungMitteRechtesDrittel - w / 2] +%[fx: $BeschriftungOffsetVertikalPixel - h / 2]" -rotate 180 \) -gravity NorthWest -composite \
    \( -fill Black -background None +size "label:$BeschriftungUntenLinks" -geometry "+%[fx: $BeschriftungMitteLinkesDrittel - w / 2] +%[fx: $BeschriftungOffsetVertikalPixel - h / 2]" -rotate 180 \) -gravity SouthWest -composite \
    \( -fill Black -background None +size "label:$BeschriftungUntenMitte" -geometry "+%[fx: $BeschriftungMitteMittleresDrittel - w / 2] +%[fx: $BeschriftungOffsetVertikalPixel - h / 2]" -rotate 180 \) -gravity SouthWest -composite \
    \( -fill Black -background None +size "label:$BeschriftungUntenRechts" -geometry "+%[fx: $BeschriftungMitteRechtesDrittel - w / 2] +%[fx: $BeschriftungOffsetVertikalPixel - h / 2]" -rotate 180 \) -gravity SouthWest -composite \
   "$DateinameErgebnis"
```

The performance is better than my previous solution with a lot of IM calls (239 sec versus 279 sec).

**Reply by viratsoft** [reply]
what is value of  $BeschriftungMitteLinkesDrittel and $BeschriftungMitteMittleresDrittel ? how can i get?

**Comment by printmaps** [comment]
> what is value of $BeschriftungMitteLinkesDrittel and $BeschriftungMitteMittleresDrittel ? how can i get?

The values are part of a script.

The basic concept is explained in snibgo's first answer.
1. create (text) image with 'label:'
2. calculate middle of image with '-geometry ...'
3. place middel of image on the desired position

**Reply by viratsoft** [reply]
invalid argument for option `-geometry': +%[fx: 1080 - w / 2] +%[fx: 1356 - h / 2]

**Reply by fmw42** [reply]
What version of ImageMagick. You can only do that with IM 7, not IM 6. Also no space between them

`-geometry "+%[fx: 1080 - w / 2]+%[fx: 1356 - h / 2]"`
